File: android/src/main/kotlin/com/example/zendesk_sdk/models/ZendeskCustomField.kt
package com.example.zendesk_sdk.models

import com.example.zendesk_sdk.ZendeskSdkChannel
import io.flutter.plugin.common.MethodCall
import zendesk.support.CustomField

/// Ticket custom field value object.
data class ZendeskCustomField(
    val fieldId: Long,
    val value: String,
) {
    fun toNative(): CustomField = CustomField(fieldId, value)

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun listFromMethodCall(call: MethodCall): List<ZendeskCustomField> {
            val customFieldsArg =
                call.argument<List<Map<String, Any>>>(ZendeskSdkChannel.Argument.CUSTOM_FIELDS)
                    ?: return emptyList()

            return customFieldsArg.mapNotNull { field ->
                val fieldId = (field[ZendeskSdkChannel.Argument.FIELD_ID] as? Number)?.toLong()
                    ?: return@mapNotNull null
                val value = field[ZendeskSdkChannel.Argument.VALUE] as? String
                    ?: return@mapNotNull null
                ZendeskCustomField(fieldId = fieldId, value = value)
            }
        }
    }
}

File: ios/zendesk_sdk/Sources/zendesk_sdk/Models/ZendeskCustomField.swift
import SupportProvidersSDK

/// Ticket custom field value object.
struct ZendeskCustomFieldModel {
    let fieldId: Int64
    let value: String

    func toNative() -> CustomField {
        CustomField(fieldId: fieldId, value: value)
    }

    static func list(from args: [String: Any]) -> [ZendeskCustomFieldModel] {
        guard let customFieldsArg = args[ZendeskSdkChannel.Argument.customFields] as? [[String: Any]] else {
            return []
        }

        return customFieldsArg.compactMap { field in
            guard let fieldIdNumber = field[ZendeskSdkChannel.Argument.fieldId] as? NSNumber,
                  let value = field[ZendeskSdkChannel.Argument.value] as? String else {
                return nil
            }
            return ZendeskCustomFieldModel(fieldId: fieldIdNumber.int64Value, value: value)
        }
    }
}
